### ISSUE TITLE
fetchpatch: ignores file renames

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#32084]
If the patch renames a file, and contains:

```
rename from libraries/base/GHC/Event/Clock.hsc
rename to libraries/base/GHC/Clock.hsc
```

Then using `patch -p1 < foobar.patch` works as expected, while using fetchpatch no rename happens.

cc @vcunat @aszlig @copumpkin @fpletz @globin

**Comment by domenkozar** [comment]
Proof that patch consideres that input: https://git.savannah.gnu.org/cgit/patch.git/commit/?h=v2.7&id=9e39d9c9950076e319d1dbbeeb7b8bd268b465f3

**Comment by nh2** [comment]
Example:

```
diff --git a/libraries/base/GHC/Event/Clock.hsc b/libraries/base/GHC/Clock.hsc
similarity index 53%
rename from libraries/base/GHC/Event/Clock.hsc
rename to libraries/base/GHC/Clock.hsc
index 5dbdb674d31..6339dc0a52b 100644
--- a/libraries/base/GHC/Event/Clock.hsc
+++ b/libraries/base/GHC/Clock.hsc
@@ -1,17 +1,25 @@
 {-# LANGUAGE Trustworthy #-}
 {-# LANGUAGE NoImplicitPrelude #-}
...
```

If in this patch, you strip everything before the `---`, then `patch` will not execute the rename.

(I personally also thought that everything before `---` was a comment and would be ignored for the purpose of applying the patch; turns out this is wrong.)

**Comment by dezgeg** [comment]
> (I personally also thought that everything before --- was a comment and would be ignored for the purpose of applying the patch; turns out this is wrong.)
The `rename` lines (like `index ...`) are a Git extension to the patch format.

**Comment by dezgeg** [comment]
Last time that was attempted, those new git-specific lines caused most existing fetchpatch hashes to change.

**Comment by aszlig** [comment]
@dezgeg: Okay, which is probably because of that very commit, because it left in other git-specific extensions. Regardless of that, I just checked to verify whether this is problem persists in 0.3.4 and even in master and it's still the case, because `lsdiff` is simply dropping renames because they're essentially non-existing hunks.

**Comment by aszlig** [comment]
Okay, while trying to patch `patchutils` it turns out that fixing the renames is a bit more involved, because the way `patchutils` does its [parsing](https://github.com/twaugh/patchutils/blob/3c70aa137ad8c4ea9085a73c0dbc1c6897b6ad45/src/filterdiff.c#L979-L1171) it's not very easy to do lookahead. Another obstacle is what @dezgeg mentioned with the changed hashes.

So what we practically need is just to turn the patch into a form that is deterministic in its results, which `patchutils` only does to a certain extent, for example it doesn't sort the file names.

Maybe it makes sense to implement what we need for `fetchpatch` ourselves and not use `patchutils`?

**Comment by dezgeg** [comment]
Yeah, I suppose writing a custom implementation is the best. After all, I think all it is doing now is stripping out any comments from the .patch files.

**Comment by globin** [comment]
We also use it to exclude files from patches and add prefixes/suffixes(?)

**Comment by aszlig** [comment]
@globin: Sure, that's no problem once we have a parser that uses proper data structures for the patches. On the other hand, however I was thinking about writing such an implementation in Haskell and parser combinators, but that would increase the closure size of `fetchpatch` a lot and also probably won't work that well because we probably have `fetchpatch` in the dependency graph of GHC (haven't checked).

So after thinking about this, I'd like to amend my comment in a way that it's still probably better to patch `fetchpatch` to use proper data structures instead of that painful `getline`-based state-all-over-the-place parser that it's currently using and also add options to it for our main use case, which is bringing the patch into a form which deterministic in its output.

**Comment by orivej** [comment]
I consider updating `patchutils` and automatically recomputing all fetchpatch hashes.

**Comment by aszlig** [comment]
@orivej: Just updating `patchutils` won't be enough, as mentioned previously because it won't handle file renames.

**Comment by arapov** [comment]
I hit this issue today. It would be nice to document this behaviour somehow .... I replaced fetchpatch() with fetchurl() and it did the trick for me.

**Comment by jtojnar** [comment]
@arapov Using `fetchurl` for dynamically generated patches (e.g. those from GitHub) is not safe since the file can change in the future. You might be better just inlining the patch into the source tree (or re-adding the renames in `fetchpatch`’s `postFetch` attribute.

**Comment by arapov** [comment]
> @arapov Using `fetchurl` for dynamically generated patches (e.g. those from GitHub) is not safe since the file can change in the future. You might be better just inlining the patch into the source tree (or re-adding the renames in `fetchpatch`’s `postFetch` attribute).

Why `fetchpatch` is not fixed the way it respects moves/renames?

**Comment by jtojnar** [comment]
@arapov We would need a `fetchpatch2` in order not to break the old hashes. And we would also need to change `patchutils` to support renames, 0.3.4 does not support it according to https://github.com/NixOS/nixpkgs/issues/32084#issuecomment-347094462, even though it is supposed to preserve some `git` metadata. Looking at the [changelogs of the newer version](https://github.com/twaugh/patchutils/releases) they did not add that either.

**Comment by bobby285271** [comment]
Still important (I think I just hit this issue, https://github.com/NixOS/nixpkgs/pull/140429#discussion_r721156003).

**Comment by eyJhb** [comment]
Still important

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/182250
1st path is by fetchpatch and 2nd is by fetchpatch2
```diff
$ diff /nix/store/namp0v9k05pjcax761h7nacpvh77205b-abc52460201bc5c7603505bb187138b0c59291aa.diff /nix/store/d2vkicv4b3xghlmaw4bnwnp01zhk7x6f-abc52460201bc5c7603505bb187138b0c59291aa.diff
0a1,2
> diff --git a/configure b/configure
> index b6c9b462f24..a86f2ceaa5b 100755
18a21,22
> diff --git a/mkspecs/common/macx.conf b/mkspecs/common/macx.conf
> index d16b77acb8e..4ba0a8eaa36 100644
28a33,37
> diff --git a/mkspecs/macx-clang/qmake.conf b/mkspecs/common/clang-macx-desktop.conf
> similarity index 83%
> rename from mkspecs/macx-clang/qmake.conf
> rename to mkspecs/common/clang-macx-desktop.conf
> index 0cf1f31b60d..042319a2aa3 100644
```

**Comment by Luflosi** [comment]
Is `fetchpatch2` supposed to fix this issue? I'm trying to apply https://github.com/teeworlds/teeworlds/commit/86d6687ef07f7f05457a7f67537b650656b13820, which only renames files and nothing else and I get the error message `error: Normalized patch '/build/patch' is empty (while the fetched file was not)!`.

**Comment by Artturin** [comment]
> Is `fetchpatch2` supposed to fix this issue? 

Yep

I'm trying to apply [teeworlds/teeworlds@86d6687](https://github.com/teeworlds/teeworlds/commit/86d6687ef07f7f05457a7f67537b650656b13820), which only renames files and nothing else and I get the error message `error: Normalized patch '/build/patch' is empty (while the fetched file was not)!`.

Experiment with https://github.com/nixos/nixpkgs/blob/41877098f3ac00295a2a9d5e5846201b969a689b/pkgs/build-support/fetchpatch/default.nix#L39 and see if you can get it working

If it can't then it's a upstream issue

**Comment by Luflosi** [comment]
I executed
```sh
nix-shell -p patchutils_0_4_2
```
then inside the shell:
```sh
wget 'https://github.com/teeworlds/teeworlds/commit/86d6687ef07f7f05457a7f67537b650656b13820.patch'
lsdiff e03553605b45c88f0b4b2980adfbbb8f6fca2fd6.patch
```
`lsdiff` outputs nothing, so this looks like an upstream issue, right. Did I invoke it correctly?

**Comment by Luflosi** [comment]
For reference, #208826 is where I wanted to use this.

**Comment by vcunat** [comment]
`fetchpatch2` still drops the rename part of https://gitlab.gnome.org/GNOME/NetworkManager-fortisslvpn/-/commit/084ef529c5fb816927ca54866f66b340265aa9f6.diff
